teps/0040-ignore-step-errors.md

```diff
@@ -145,13 +156,294 @@ This proposal is limited to a step within a task and does not address `pipelineT
+    // set onError to fail to indicate the entrypoint to exit the taskRun if the container exits with non zero exit code
+    // set onError to continue to indicate the entrypoint to continue executing the rest of the steps irrespective of the container exit code
+    OnError string `json:"onError,omitempty"`
+}
+```
+
+A `task` author or a `pipeline` author can use `onError` to ignore the step error. If `onError` is
+set to `continue`, the entrypoint sets the original failed exit code of the `script` or the wrapped command in the
+container termination state.
+
+To ignore a step error, set the `onError` to `continue`:
+
+```yaml
+steps:
+  - image: docker.io/library/golang:latest
+    name: ignore-unit-test-failure
+    onError: continue
+    script: |
+      go test .
+```
+
+A `step` with `onError` set to `continue` does not fail the `taskRun` and continues executing the rest of the steps in
+a task. The original failed exit code of the wrapped command is available in the termination state of the container.
+
+```
+kubectl get tr taskrun-unit-test-t6qcl -o json | jq .status
+{
+  "completionTime": "2021-06-21T18:22:06Z",
+  "conditions": [
+    {
+      "lastTransitionTime": "2021-06-21T18:22:06Z",
+      "message": "All Steps have completed executing",
+      "reason": "Succeeded",
+      "status": "True",
+      "type": "Succeeded"
+    }
+  ],
+  "podName": "taskrun-unit-test-t6qcl-pod-zpqs9",
+  "startTime": "2021-06-21T18:21:57Z",
+  "steps": [
+    {
+      "container": "step-ignore-unit-test-failure",
+      "imageID": "...",
+      "name": "ignore-unit-test-failure",
+      "terminated": {
+        "containerID": "...",
+        "exitCode": 1,
+        "finishedAt": "2021-06-21T18:22:05Z",
+        "reason": "Completed",
+        "startedAt": "2021-06-21T18:22:05Z"
+      }
+    },
+  ],
+```
+
+As part of this design, we are introducing an additional internal volume `/tekton/steps/`. A file named `exitCode` will
+be created for each step to store the non-zero exit code, for example:
+
+```
+/tekton/steps/<step-name>/exitCode
+```
+
+`<step-name>` will be replaced with the name of the step. If a step does not have any name, `<step-name>` will be
+replaced with `step-unnamed-<step-index>` where `<step-index>` is `0` for the first step, `1` for the second step, so
+on and so forth.
+
+This new internal volume `/tekton/steps/<step-name>/` can be utilized in future to collect any metadata for each step.
+And instead of referencing to it with the `<step-name>`, will create a symlink such that it can be accessed using the
+`<step-index>`.
+
+```shell
+ln -s /tekton/steps/<step-name> /tekton/steps/<step-index>
+```
+
+Any subsequent step can access the non-zero exit code of a previous step using the `path` similar to a task result,
+for example:
+
+```shell
+$(steps.<step-name>.exitCode.path)
+```
+
+The `exitCode` of a step without any name can be referenced using:
+
+```shell
+$(steps.step-unnamed-<step-index>.exitCode.path)
```

Comment:
NIT: step-unnamed-1 could also be a step name (not a very likely one :P) - it's probably enough to doc that that is not a valid step name to use :)